tutorial/source/svi_flow_guide.ipynb

```diff
@@ -0,0 +1,236 @@
+    "from pyro.infer import SVI, Trace_ELBO\n",
+    "from torch import Tensor"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Model\n",
+    "\n",
+    "We define a simple non-linear model $p(x | z)$ with a standard Gaussian prior $p(z)$ over the latent variables $z$."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "prior = pyro.distributions.Normal(torch.zeros(3), torch.ones(3)).to_event(1)\n",
+    "\n",
+    "def likelihood(z: Tensor):\n",
+    "    mu = z[..., :2]\n",
+    "    rho = z[..., 2].tanh() * 0.99\n",
+    "\n",
+    "    cov = 1e-2 * torch.stack([\n",
+    "        torch.ones_like(rho), rho,\n",
+    "        rho, torch.ones_like(rho),\n",
+    "    ], dim=-1).unflatten(-1, (2, 2))\n",
+    "\n",
+    "    return pyro.distributions.MultivariateNormal(mu, cov)\n",
+    "\n",
+    "def model(x: Tensor):\n",
+    "    with pyro.plate(\"batch\", len(x)):\n",
+    "        z = pyro.sample(\"z\", prior)\n",
+    "\n",
+    "        with pyro.plate(\"obs\", 5):\n",
+    "            pyro.sample(\"x\", likelihood(z), obs=x.transpose(0, 1))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We sample 64 reference latent variables and observations $(z^*, x^*)$. In practice, $z^*$ is unknown, and $x^*$ is your data."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "z_star = prior.sample((64,))\n",
+    "x_star = likelihood(z_star).sample((5,)).transpose(0, 1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Guide\n",
+    "\n",
+    "We define the guide $q_\\phi(z | x)$ with a normalizing flow. We choose a conditional [neural spline flow](https://arxiv.org/abs/1906.04032) borrowed from the the [Zuko](https://zuko.readthedocs.io/) library. Because Zuko distributions are very similar to Pyro distributions, a thin wrapper (`Zuko2Pyro`) is sufficient to make Zuko and Pyro 100% compatible."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "flow = zuko.flows.NSF(features=3, context=10, transforms=1, hidden_features=(256, 256))\n",
+    "flow.transform = flow.transform.inv  # inverse autoregressive flow (IAF) are fast to sample from\n",
+    "\n",
+    "def guide(x: Tensor):\n",
+    "    pyro.module(\"flow\", flow)\n",
+    "\n",
+    "    with pyro.plate(\"batch\", len(x)):  # amortized\n",
+    "        pyro.sample(\"z\", Zuko2Pyro(flow(x.flatten(-2))))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## SVI\n",
+    "\n",
+    "We train our guide with a standard stochastic variational inference (SVI) pipeline. We use 16 particles to reduce the variance of the ELBO and clip the norm of the gradients to make training more stable."
```

Comment:
quite aggressive clipping?

Reply:
That's a value I use quite often, but I don't have much intuition about it.